pkg/operator/daemonsetstatuscontroller/daemonsetstatuscontroller.go

```diff
@@ -0,0 +1,226 @@
+
+	return c
+}
+
+func (c *DaemonSetStatusController) sync() error {
+	actualDaemonSet, err := c.daemonSetLister.DaemonSets(c.targetNamespace).Get(c.targetName)
+	if err != nil {
+		return err
+	}
+
+	errors := []error{}
+	if actualDaemonSet == nil {
+		message := fmt.Sprintf("daemonset.v1.apps/%v: could not be retrieved", c.targetName)
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "APIServerDaemonSetAvailable",
+				Status:  operatorv1.ConditionFalse,
+				Reason:  "NoDaemon",
+				Message: message,
+			})))...,
+		)
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "APIServerDaemonSetProgressing",
+				Status:  operatorv1.ConditionTrue,
+				Reason:  "NoDaemon",
+				Message: message,
+			})))...,
+		)
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "APIServerDaemonSetDegraded",
+				Status:  operatorv1.ConditionTrue,
+				Reason:  "NoDaemon",
+				Message: message,
+			})))...,
+		)
+
+		return utilerrors.NewAggregate(errors)
+	}
+
+	if actualDaemonSet.Status.NumberAvailable == 0 {
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "APIServerDaemonSetAvailable",
+				Status:  operatorv1.ConditionFalse,
+				Reason:  "NoAPIServerPod",
+				Message: "no openshift-apiserver daemon pods available on any node.",
+			})))...,
+		)
+	} else {
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:   "APIServerDaemonSetAvailable",
+				Status: operatorv1.ConditionTrue,
+				Reason: "AsExpected",
+			})))...,
+		)
+	}
+
+	// If the daemonset is up to date and the operatorConfig are up to date, then we are no longer progressing
+	daemonSetAtHighestGeneration := actualDaemonSet.ObjectMeta.Generation == actualDaemonSet.Status.ObservedGeneration
+	if !daemonSetAtHighestGeneration {
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "APIServerDaemonSetProgressing",
+				Status:  operatorv1.ConditionTrue,
+				Reason:  "NewGeneration",
+				Message: fmt.Sprintf("daemonset/apiserver.openshift-operator: observed generation is %d, desired generation is %d.", actualDaemonSet.Status.ObservedGeneration, actualDaemonSet.ObjectMeta.Generation),
+			})))...,
+		)
+	} else {
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:   "APIServerDaemonSetProgressing",
+				Status: operatorv1.ConditionFalse,
```

Comment:
I do not think this is sufficient for Progressing=False.  We should [also][1] look for `numberReady == currentNumberScheduled` and `updatedNumberScheduled == currentNumberScheduled`.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1792382#c2

Reply:
> I do not think this is sufficient for Progressing=False. We should [also](https://bugzilla.redhat.com/show_bug.cgi?id=1792382#c2) look for `numberReady == currentNumberScheduled` and `updatedNumberScheduled == currentNumberScheduled`.

If daemonsets are tripping their generations right, then this is fine.  We aren't progressing.  If something is unavailable (don't use the ready condition), it will be degraded.

Reply:
> If something is unavailable (don't use the ready condition), it will be degraded.

If our desired state differs from our current state, I think we are attempting to reconcile the difference (e.g. by crashlooping the pod), and we are still attempting to progress when that is going on, even if we aren't actually making progress.  The [current API docs][1] aren't really clear on whether Progressing=True is for "state does not match our desired spec" or for the more narrow "... and we continue to make useful progress towards fixing that" (vs. "... and we just keep banging our heads into the same wall; no idea if we'll ever get there").  I'd thought it was just state != spec; you seem to feel it's "... but we're getting closer".  Can we land some doc language so it's clear to the CVO and other operator authors what the intended semantics are?

[1]: https://github.com/openshift/api/blob/e7ccdda6ba67b868baceb40b6db943355967f26d/config/v1/types_cluster_operator.go#L146-L148

Reply:
> clear on whether Progressing=True is for "state does not match our desired spec" or for the more narrow "... and we continue to make useful progress towards fixing that" (vs. "... and we just keep banging our heads into the same wall; no idea if we'll ever get there")

It is unreasonable to expect an operator author to be able to distinguish failures that will eventually resolve from those that won't.  There are a few narrow cases where it is possible and there are a multiplicity of unknown errors where it is not possible to know.

In this particular case, we have a tactical, narrow fix to alleviate pressure in a previous release and a far more complete change going into 4.4 that is too risky to backport: https://github.com/openshift/cluster-openshift-apiserver-operator/pull/303 .  I wouldn't devote more effort to the "edges of edges" handling for code that *will never exist in master*.

If the generation/status.generation of daemonsets doesn't work as expected, I'm open to more changes/niggling, but if it does, this *throw-away code* is good enough to alleviate our issues.